File: models/photo.go
package models

import (
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type WorkFlow struct {
	UpvoteGrade   int    `json:"upvote_grade" bson:"upvote_grade,omitempty"`
	WorkflowStage string `json:"workflow_stage" bson:"workflow_stage,omitempty"`
}

type PhotoMetadata struct {
	ShotDate     int64       `json:"shot_date" bson:"shot_date"`
	ModifiedDate int64       `json:"process_date" bson:"process_date"`
	Camera       string      `json:"camera" bson:"camera"`
	Location     Geolocation `json:"location" bson:"location"`
	Place        Place       `json:"place" bson:"place"`
	Exposure     string      `json:"exposure" bson:"exposure"`
	FNumber      float32     `json:"f_number" bson:"f_number"`
	ISO          int         `json:"iso" bson:"iso"`
	Description  string      `json:"description" bson:"description"`
}

type Comments struct {
	CommenterID   primitive.ObjectID `json:"commenter_id" bson:"commenter_id"`
	CommenterName string             `json:"commenter_name" bson:"commenter_name"`
	Comment       string             `json:"comment" bson:"comment"`
	Time          int64              `json:"time" bson:"time"`
}

type PhotoModel struct {
	ID        primitive.ObjectID   `json:"id" bson:"id"`
	Url       string               `json:"url" bson:"url"`
	IsPublic  bool                 `json:"is_public" bson:"is_public"`
	Owner     primitive.ObjectID   `json:"owner" bson:"owner"`
	VisibleTo []primitive.ObjectID `json:"visible_to" bson:"visible_to"`
	Metadata  PhotoMetadata        `json:"metadata" bson:"metadata"`
	WorkFlow  WorkFlow             `json:"workflow" bson:"workflow_stage"`
	SimilarTo []primitive.ObjectID `json:"similar_to" bson:"similar_to"`
	Ancestor  primitive.ObjectID   `json:"ancestor" bson:"ancestor"`
	Comments  []Comments           `json:"comments" bson:"comments"`
}

type PhotoSearchOptions struct {
	ShotAfter      *int64  `json:"shot_after,omitempty"`
	ShotBefore     *int64  `json:"shot_before,omitempty"`
	ModifiedAround *int64  `json:"modified_around,omitempty"`
	Camera         *string `json:"camera,omitempty"`
	Location       *struct {
		Geolocation Geolocation `json:"geolocation,omitempty"`
		Radius      float64     `json:"radius,omitempty"`
	} `json:"location,omitempty"`
	LocationContains   *string `json:"location_contains,omitempty"`
	CommentsContaining *string `json:"comments_containing,omitempty"`
	OwnedPhotoFilter   *struct {
		OnlyMine      *bool   `json:"only_mine,omitempty"`
		IsPublic      *bool   `json:"is_public,omitempty"`
		UpvoteGrade   *int8   `json:"upvote_grade,omitempty"`
		WorkflowStage *string `json:"workflow_stage,omitempty"`
	} `json:"owned_photo_filter,omitempty"`
}

File: Client/Android/app/src/main/java/com/photomanager/photomanager/main/home/api/model/Photo.kt
package com.photomanager.photomanager.main.home.api.model

import com.photomanager.photomanager.main.home.api.Geolocation
import com.photomanager.photomanager.main.home.api.ObjectId
import com.photomanager.photomanager.main.home.api.Place
import com.photomanager.photomanager.main.home.api.typeconverters.WorkflowStageSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable


@Serializable(with = WorkflowStageSerializer::class)
enum class WorkflowStage(val value: String) {
   FOOTAGE("footage"),
   COLLECTION("collection"),
   ALBUM("album")
}

@Serializable
data class WorkFlow(
   @SerialName("upvote_grade") val upvoteGrade: Int,
   @SerialName("workflow_stage") val workflowStage: WorkflowStage,
   @SerialName("albums") val albums: List<ObjectId>
)

@Serializable
data class PhotoMetadata(
   @SerialName("shot_date") val shotDate: Long?,
   @SerialName("process_date") val modifiedDate: Long?,
   @SerialName("camera") val camera: String?,
   @SerialName("location") val location: Geolocation?,
   @SerialName("place") val place: Place?,
   @SerialName("exposure") val exposure: String?,
   @SerialName("f_number") val fNumber: Float?,
   @SerialName("iso") val iso: Int?,
   @SerialName("description") val description: String?
)

@Serializable
data class Comments(
   @SerialName("commenter_id") val commenterID: ObjectId,
   @SerialName("commenter_name") val commenterName: String,
   @SerialName("comment") val comment: String,
   @SerialName("time") val time: Long
)

@Serializable
data class PhotoModel(
   @SerialName("id") val id: ObjectId,
   @SerialName("url") val url: String,
   @SerialName("is_public") val isPublic: Boolean,
   @SerialName("owner") val owner: ObjectId,
   @SerialName("visible_to") val visibleTo: List<ObjectId>,
   @SerialName("metadata") val metadata: PhotoMetadata,
   @SerialName("workflow") val workFlow: WorkFlow,
   @SerialName("similar_to") val similarTo: List<ObjectId>,
   @SerialName("ancestor") val ancestor: ObjectId,
   @SerialName("comments") val comments: List<Comments>,
   @SerialName("tags") val tags: List<String>
)

@Serializable
data class PhotoSearchLocation(
   @SerialName("geolocation") val geolocation: Geolocation,
   @SerialName("radius") val radius: Double
)

@Serializable
data class UpvoteGradeRange(
   @SerialName("min") val min: Int,
   @SerialName("max") val max: Int
)

@Serializable
data class PhotoSearchOwnedPhotoFilter(
   @SerialName("only_mine") val onlyMine: Boolean?,
   @SerialName("is_public") val isPublic: Boolean?,
   @SerialName("upvote_grade") val upvoteGrade: UpvoteGradeRange?,
   @SerialName("workflow_stage") val workflowStage: WorkflowStage?
)

@Serializable
data class PhotoSearchOptions(
   @SerialName("shot_after") val shotAfter: Long?,
   @SerialName("shot_before") val shotBefore: Long?,
   @SerialName("modified_around") val modifiedAround: Long?,
   @SerialName("camera") val camera: String?,
   @SerialName("location") val location: PhotoSearchLocation?,
   @SerialName("location_contains") val locationContains: String?,
   @SerialName("comments_containing") val commentsContaining: String?,
   @SerialName("owned_photo_filter") val ownedPhotoFilter: PhotoSearchOwnedPhotoFilter?
)

@Serializable
data class AlbumSearchCriteria(
   @SerialName("owner_id") val ownerID: ObjectId?,
   @SerialName("name") val nameRegex: String?,
   @SerialName("description") val descriptionRegex: String?,
   @SerialName("visibility_to") val visibilityTo: ObjectId?
)